### ISSUE TITLE
Poor baselines on CIFAR-10 for flax_resnet.py example

### ISSUE THREAD

**Issue by fabianp** [google/jaxopt#156]
standard baselines are able to reach 90+ accuracy without much tweaking (see for example https://github.com/kuangliu/pytorch-cifar ), while our example never goes beyond 70% accuracy on the validation set

**Comment by fabianp** [comment]
taking this on during the sprint

**Comment by fabianp** [comment]
Update: I wrote an alternative to flax_resnet.py that uses GroupNorm instead of BatchNorm: https://gist.github.com/fabianp/7c6c2da33156010d5d822bdad997c942

It doesn't seem to do better on the test set accuracy, which is stuck around 70%, although train accuracy is at 95%

**Comment by fabianp** [comment]
CC @fllinares who is facing similar issues

**Comment by fabianp** [comment]
This is the train/test accuracy on CIFAR10 with the latest script with GroupNorm ( https://gist.github.com/fabianp/7c6c2da33156010d5d822bdad997c942)

![Figure_1](https://user-images.githubusercontent.com/277639/152154195-c67759e1-15ac-44a2-bb6d-446a91ec7d33.png)

**Comment by ita9naiwa** [comment]
When I was playing with this lib for the first time, test accuracy is too high `0.99`.

I ran the same code again `flax_resnet.py` and get the same results.
I saw no 'validation dataset' so I'm not sure what I'm talking about now is your concern but

----
I'm using Jax==0.2.28, jaxlib==0.1.76, and main branch of jaxopt

```
(base) vnode273:~/jaxopt/examples/deep_learning# python flax_resnet.py 
I0202 21:39:20.349320 139792832852928 dataset_info.py:439] Load dataset info from /root/tensorflow_datasets/mnist/3.0.1
I0202 21:39:20.351258 139792832852928 dataset_info.py:492] Field info.citation from disk and from code do not match. Keeping the one from code.
I0202 21:39:20.351508 139792832852928 dataset_info.py:492] Field info.splits from disk and from code do not match. Keeping the one from code.
I0202 21:39:20.351602 139792832852928 dataset_info.py:492] Field info.supervised_keys from disk and from code do not match. Keeping the one from code.
I0202 21:39:20.351672 139792832852928 dataset_info.py:492] Field info.module_name from disk and from code do not match. Keeping the one from code.
I0202 21:39:20.351815 139792832852928 dataset_builder.py:369] Reusing dataset mnist (/root/tensorflow_datasets/mnist/3.0.1)
I0202 21:39:20.351910 139792832852928 logging_logger.py:44] Constructing tf.data.Dataset mnist for split train, from /root/tensorflow_datasets/mnist/3.0.1
I0202 21:39:20.413569 139792832852928 dataset_info.py:439] Load dataset info from /root/tensorflow_datasets/mnist/3.0.1
I0202 21:39:20.414881 139792832852928 dataset_info.py:492] Field info.citation from disk and from code do not match. Keeping the one from code.
I0202 21:39:20.415047 139792832852928 dataset_info.py:492] Field info.splits from disk and from code do not match. Keeping the one from code.
I0202 21:39:20.415124 139792832852928 dataset_info.py:492] Field info.supervised_keys from disk and from code do not match. Keeping the one from code.
I0202 21:39:20.415191 139792832852928 dataset_info.py:492] Field info.module_name from disk and from code do not match. Keeping the one from code.
I0202 21:39:20.415324 139792832852928 dataset_builder.py:369] Reusing dataset mnist (/root/tensorflow_datasets/mnist/3.0.1)
I0202 21:39:20.415408 139792832852928 logging_logger.py:44] Constructing tf.data.Dataset mnist for split test, from /root/tensorflow_datasets/mnist/3.0.1
I0202 21:39:20.457857 139792832852928 xla_bridge.py:247] Unable to initialize backend 'tpu_driver': NOT_FOUND: Unable to find driver in registry given worker: 
I0202 21:39:20.680176 139792832852928 xla_bridge.py:247] Unable to initialize backend 'tpu': INVALID_ARGUMENT: TpuPlatform is not available.
2022-02-02 21:39:41.326129: E external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_dnn.cc:5205] Disabling cuDNN frontend for the following convolution:
  input: {count: 256 feature_map_count: 256 spatial: 3 3  value_min: 0.000000 value_max: 0.000000 layout: BatchDepthYX}
  filter: {output_feature_map_count: 256 input_feature_map_count: 256 layout: OutputInputYX shape: 2 2 }
  {zero_padding: 0 0  pad_alignment: default filter_strides: 1 1  dilation_rates: 1 1 }
  ... because it uses an identity activation.
2022-02-02 21:39:41.330710: E external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_dnn.cc:5205] Disabling cuDNN frontend for the following convolution:
  input: {count: 256 feature_map_count: 256 spatial: 2 2  value_min: 0.000000 value_max: 0.000000 layout: BatchDepthYX}
  filter: {output_feature_map_count: 256 input_feature_map_count: 256 layout: OutputInputYX shape: 2 2 }
  {zero_padding: 1 1  pad_alignment: default filter_strides: 1 1  dilation_rates: 1 1 }
  ... because it uses an identity activation.
2022-02-02 21:39:41.336629: E external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_dnn.cc:5205] Disabling cuDNN frontend for the following convolution:
  input: {count: 512 feature_map_count: 256 spatial: 1 1  value_min: 0.000000 value_max: 0.000000 layout: BatchDepthYX}
  filter: {output_feature_map_count: 512 input_feature_map_count: 256 layout: OutputInputYX shape: 1 1 }
  {zero_padding: 1 1  pad_alignment: default filter_strides: 1 1  dilation_rates: 1 1 }
  ... because it uses an identity activation.
2022-02-02 21:39:41.370893: E external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_dnn.cc:5205] Disabling cuDNN frontend for the following convolution:
  input: {count: 512 feature_map_count: 512 spatial: 1 1  value_min: 0.000000 value_max: 0.000000 layout: BatchDepthYX}
  filter: {output_feature_map_count: 256 input_feature_map_count: 512 layout: OutputInputYX shape: 1 1 }
  {zero_padding: 0 0  pad_alignment: default filter_strides: 1 1  dilation_rates: 1 1 }
  ... because it uses an identity activation.
2022-02-02 21:39:41.373978: E external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_dnn.cc:5205] Disabling cuDNN frontend for the following convolution:
  input: {count: 256 feature_map_count: 512 spatial: 2 2  value_min: 0.000000 value_max: 0.000000 layout: BatchDepthYX}
  filter: {output_feature_map_count: 256 input_feature_map_count: 512 layout: OutputInputYX shape: 1 1 }
  {zero_padding: 0 0  pad_alignment: default filter_strides: 1 1  dilation_rates: 1 1 }
  ... because it uses an identity activation.
2022-02-02 21:39:41.378141: E external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_dnn.cc:5205] Disabling cuDNN frontend for the following convolution:
  input: {count: 128 feature_map_count: 256 spatial: 6 6  value_min: 0.000000 value_max: 0.000000 layout: BatchDepthYX}
  filter: {output_feature_map_count: 256 input_feature_map_count: 256 layout: OutputInputYX shape: 3 3 }
  {zero_padding: 0 0  pad_alignment: default filter_strides: 1 1  dilation_rates: 1 1 }
  ... because it uses an identity activation.
[Epoch 0/10] Train acc: 0.977, train loss: 0.449. Test acc: 0.981, test loss: 0.439. Time elapsed: 0:00:46
[Epoch 1/10] Train acc: 0.992, train loss: 0.382. Test acc: 0.986, test loss: 0.394. Time elapsed: 0:00:54
[Epoch 2/10] Train acc: 0.980, train loss: 0.358. Test acc: 0.988, test loss: 0.360. Time elapsed: 0:01:02
[Epoch 3/10] Train acc: 1.000, train loss: 0.302. Test acc: 0.989, test loss: 0.330. Time elapsed: 0:01:10
[Epoch 4/10] Train acc: 0.980, train loss: 0.356. Test acc: 0.987, test loss: 0.319. Time elapsed: 0:01:18
[Epoch 5/10] Train acc: 0.980, train loss: 0.291. Test acc: 0.983, test loss: 0.304. Time elapsed: 0:01:26
[Epoch 6/10] Train acc: 0.992, train loss: 0.245. Test acc: 0.990, test loss: 0.264. Time elapsed: 0:01:34
[Epoch 7/10] Train acc: 0.996, train loss: 0.224. Test acc: 0.990, test loss: 0.245. Time elapsed: 0:01:42
[Epoch 8/10] Train acc: 0.996, train loss: 0.206. Test acc: 0.991, test loss: 0.222. Time elapsed: 0:01:50
[Epoch 9/10] Train acc: 0.996, train loss: 0.188. Test acc: 0.990, test loss: 0.214. Time elapsed: 0:01:58
```

**Comment by fabianp** [comment]
thanks @ita9naiwa for looking into it. I think you might be running on MNIST, where accuracy is indeed at 99%. 

Can you run it in the cifar dataset, for this add --dataset=cifar10 to the command line, something like:
    python flax_resnet.py --dataset=cifar10

**Comment by fabianp** [comment]
BTW by validation accuracy I mean the "Test acc" reported

**Comment by ita9naiwa** [comment]
@fabianp it seems that normalization is missing in `flax_resnet.py`.

**Comment by ita9naiwa** [comment]
sorry, I found normalization in function `prediction`
https://github.com/google/jaxopt/blob/1f4ebbd705d2557af4d4929db2b5088b8ce3426b/examples/deep_learning/flax_resnet.py#L168

**Comment by ita9naiwa** [comment]
@fabianp
I ran `python flax_resnet.py --dataset=cifar10` and found similar results with you.

```
[Epoch 0/10] Train acc: 0.457, train loss: 2.016. Test acc: 0.472, test loss: 1.874. Time elapsed: 0:00:46
[Epoch 1/10] Train acc: 0.609, train loss: 1.515. Test acc: 0.558, test loss: 1.656. Time elapsed: 0:00:54
[Epoch 2/10] Train acc: 0.648, train loss: 1.425. Test acc: 0.590, test loss: 1.591. Time elapsed: 0:01:01
[Epoch 3/10] Train acc: 0.684, train loss: 1.243. Test acc: 0.637, test loss: 1.466. Time elapsed: 0:01:09
[Epoch 4/10] Train acc: 0.719, train loss: 1.155. Test acc: 0.655, test loss: 1.411. Time elapsed: 0:01:17
[Epoch 5/10] Train acc: 0.777, train loss: 0.940. Test acc: 0.689, test loss: 1.335. Time elapsed: 0:01:24
[Epoch 6/10] Train acc: 0.840, train loss: 0.839. Test acc: 0.707, test loss: 1.326. Time elapsed: 0:01:32
[Epoch 7/10] Train acc: 0.844, train loss: 0.780. Test acc: 0.713, test loss: 1.360. Time elapsed: 0:01:40
[Epoch 8/10] Train acc: 0.887, train loss: 0.671. Test acc: 0.718, test loss: 1.380. Time elapsed: 0:01:48
[Epoch 9/10] Train acc: 0.824, train loss: 0.916. Test acc: 0.688, test loss: 1.550. Time elapsed: 0:01:55
```

**Comment by ita9naiwa** [comment]
I guess learning rate without learning rate scheduler makes model cannot converge.

adding simple scheduler gives test acc. >= 0.8

```python
pol_scale = optax.polynomial_schedule(init_value=1.0, end_value=0.001, power=1, transition_steps=5000)

  opt = optax.chain(optax.sgd(learning_rate=FLAGS.learning_rate,
                  momentum=FLAGS.momentum,
                  nesterov=True),
                  optax.scale_by_schedule(pol_scale))
```

**Comment by fabianp** [comment]
There are still some things I don't understand. For example, this code (https://github.com/fabianp/pytorch-cifar) achieves 80% test set accuracy after just 5 epochs in Pytorch using constant-learning rate Adam and without any data augmentation.

In jaxopt, with the same parameters, the test set accuracy never goes above 70%

**Comment by ita9naiwa** [comment]
@fabianp I ran `https://github.com/fabianp/pytorch-cifar` and it's suspicious that training loss does not go down and stops around 0.3 in `flax_resnet.py` unlike pytorch example. (but it might be because we put regularization loss in jaxopt)

I also guess that one possible (but critical) cause for this is that the training dataset is not properly shuffled in `flax_resnet.py`

**Comment by ita9naiwa** [comment]
@fabianp I was thinking about this and found some possible reasons for this.

1. diff between micro/macro average
https://datascience.stackexchange.com/questions/15989/micro-average-vs-macro-average-performance-in-a-multiclass-classification-settin

pytorch example uses macro average (sum all, then mean), but flax example here uses micro (means of means)

2. `test_acc, test_loss = 0., 0.`
it affects badly,maybe somewhat better like (even quite verbose)
```python
test_acc, test_loss = None, None
      # make a pass over test set to compute test accuracy
      for _ in range(iter_per_epoch_test):
          tmp = accuracy_and_loss(params, FLAGS.l2reg, next(test_ds), batch_stats)
          if test_acc is not None:
            test_acc += tmp[0] / iter_per_epoch_test
            test_loss += tmp[1] / iter_per_epoch_test
          else:
            test_acc = tmp[0] / iter_per_epoch_test
            test_loss = tmp[1] / iter_per_epoch_test
```

3. as I mentioned before, 
https://www.tensorflow.org/api_docs/python/tf/data/Dataset#shuffle

it seems that `tfds.shuffle` only shuffle within the buffer(buffer size given) thus,

https://github.com/google/jaxopt/blob/eb6e75dfee1d25cc2b206ad1410668c576bf6750/examples/deep_learning/flax_resnet.py#L68-L70
this line might change to work well.

**Comment by fabianp** [comment]
Thanks a lot @ita9naiwa , this is extremely helpful. I've been busy for the last couple of days but will get back to this and try these suggestions out as soon as possible :-)

**Comment by fabianp** [comment]
Coming back to this issue.  I don't think it's a dataset loading issue. Here I adapted the current example to load the dataset using pytorch, including data augmentation that yields good generalization accuracy (https://gist.github.com/fabianp/3ead88c85bc43836dde19cc425d5e81b) and it exhibits the same problem

As for the micro/macro average, I think this refers to averages computing across classes, while here the micro averages are across mini-batches. If $a_i$ is the accuracy in the $i$-th mini-batch and $n_b$ the number of samples in the mini-batch, we're computing
$$
\frac{n_b}{n} \sum_i \frac{1}{n_b} a_i = \frac{1}{n} \sym_i a_i
$$
which is the same as computing the accuracy in the full dataset and then averaging (modulo truncation of the last mini-batch). Let me know if I got it wrong, or if you tried and get different results.

**Comment by fabianp** [comment]
Fixed 1.5 years later in https://github.com/google/jaxopt/commit/75724b574fe95965bf908e8aa14fe2ced4eead67 🎉
